doc/configuration.rst
Configuration
=============

.. contents::
    :local:
    :depth: 2

You can specify a configuration file by exporting the
``UDATA_SETTINGS`` environment variable.

.. code-block:: shell

    export UDATA_SETTINGS=/path/to/my/udata.cfg


The configuration file is simply a Python file exporting some variables.


Example configuration file
--------------------------

Here a sample configuration file:

.. code-block:: python

    # -*- coding: utf-8 -*-
    from __future__ import unicode_literals


    DEBUG = True
    ASSETS_DEBUG = True
    ASSETS_AUTO_BUILD = True
    # REQUIREJS_RUN_IN_DEBUG = True

    SEND_MAIL = False

    SECRET_KEY = 'A unique secret key'

    USE_SSL = True
    SERVER_NAME = 'www.data.dev'

    DEFAULT_LANGUAGE = 'fr'
    PLUGINS =  'gouvfr piwik harvest me'.split()

    SITE_ID = 'www.data.dev'
    SITE_TITLE = 'Data.dev'
    SITE_URL = 'www.data.dev'

    DEBUG_TOOLBAR = True

    FS_PREFIX = '/s'
    FS_ROOT = '/srv/http/www.data.dev/fs'


Availables options
------------------

uData use some flask extensions and so provides all avaiable options for theses extensions.
Most of the time, it tries to provide some sane defaults.


Flask and global behavior options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

DEBUG
*****

**default**: ``False``

A boolean specifying the debug mode.

SEND_MAIL
*********

**default**: ``True``

A boolean specifying if the emails should actually be sent.

USE_SSL
*******

**default**: ``False``

A boolean used to force SSL usage for logged in users.


DEFAULT_LANGUAGE
****************

**default**: ``'en'``

The default language used when no language prefix is provided in URLs.


SECRET_KEY
**********

A secret key used as salt for cryptographic parts.
**You must specify your own secure key and use same in all your instances**

SITE_ID
*******

**default**: ``'default'``

The site identifier. It is used to ataached some database configuration, metrics...


PLUGINS
*******

**default**: ``[]``

A list of enabled uData plugins.


THEME
*****

**default**: ``'default'``


The enabled theme name.


TEMPLATE_CACHE_DURATION
***********************

**default**: ``5``

The duration used for templates cache, in minutes.


Flask-Assets options
~~~~~~~~~~~~~~~~~~~~

ASSETS_DEBUG
************

**default**: ``False``

A boolean indicating whether assets should compiled or not allowing in-browser debugging.


ElasticSearch configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~

ELASTICSEARCH_URL
*****************

**default**: ``'localhost:9200'``

The elasticsearch server url used for search indexing.

.. code-block:: python

    ELASTICSEARCH_URL = 'elasticserver:9200'


Mongoengine/Flask-Mongoengine options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

MONGODB_HOST
************

**default**: ``localhost``

The mongodb hostname used by udata.


MONGODB_PORT
************

**default**: ``27017``

The mongodb post used by udata.

MONGODB_DB
**********

**default**: ``udata``

The mongodb database used by udata. When testing, the test database will use the same name suffixed by ``-test``


See `the official Flask-MongoEngine documentation <https://flask-mongoengine.readthedocs.org/en/latest/>`_ for more details.


Celery options
~~~~~~~~~~~~~~

By default, uData is configured to use Redis as Celery backend and a cutomized MongoDB scheduler.

The defaults are:

.. code-block:: python

    BROKER_URL = 'redis://localhost:6379'
    BROKER_TRANSPORT_OPTIONS = {
        'fanout_prefix': True,
        'fanout_patterns': True,
    }
    CELERY_RESULT_BACKEND = 'redis://localhost:6379'
    CELERY_ACCEPT_CONTENT = ['pickle', 'json']
    CELERYD_HIJACK_ROOT_LOGGER = False
    CELERYBEAT_SCHEDULER = 'udata.tasks.Scheduler'
    CELERY_MONGODB_SCHEDULER_COLLECTION = "schedules"

You can see the full list of Celery options in the
`Celery official documentation <http://docs.celeryproject.org/en/latest/configuration.html>`_.


Flask-Mail options
~~~~~~~~~~~~~~~~~~

You can see the full configuration option list in
`the official Flask-Mail documentation <https://pythonhosted.org/flask-mail/>`_

MAIL_DEFAULT_SENDER
*******************

**default**: ``'webmaster@udata'``

The default identity used for outgoing mails.


Flask-OAuthlib options
~~~~~~~~~~~~~~~~~~~~~~

uData is Oauthlib to provide OAuth2 on the API.
The full option list is available in
`the official Flask-OAuthlib documentation <https://flask-oauthlib.readthedocs.org/en/latest/oauth2.html#configuration>`_

OAUTH2_PROVIDER_TOKEN_EXPIRES_IN
********************************

**default**: ``30 * 24 * 60 * 60`` (30 days)

The OAuth2 token duration.


OAUTH2_PROVIDER_ERROR_ENDPOINT
******************************

**default**: ``'oauth-i18n.oauth_error'``

The OAuth2 error page. Do not modify unless you know what you do.


Flask-Cache options
~~~~~~~~~~~~~~~~~~~

uData use Flask-Cache to handle cache and use Redis by default.
You can see the full options list in
`the official Flask-Cache documentation <https://pythonhosted.org/Flask-Cache/>`_

CACHE_TYPE
**********

**default**: ``'redis'``

The cache type, which can be adjusted to your needs.


CACHE_KEY_PREFIX
****************

**default**: ``'udata-cache'``

A prefix used for cache keys to avoid conflict with other middleware.
It also allows you use the same backend with different instances.


Flask-FS options
~~~~~~~~~~~~~~~~

uData use Flask-FS as storage abstraction.


udata/settings.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals


class Defaults(object):
    DEBUG = False
    TESTING = False
    SEND_MAIL = True
    LANGUAGES = {
        'en': 'English',
        'fr': 'Français',
        'es': 'Español',
    }
    DEFAULT_LANGUAGE = 'en'
    SECRET_KEY = 'Default uData secret key'

    MONGODB_HOST = 'localhost'
    MONGODB_PORT = 27017
    MONGODB_DB = 'udata'

    # BROKER_TRANSPORT = 'redis'
    BROKER_URL = 'redis://localhost:6379'
    BROKER_TRANSPORT_OPTIONS = {
        'fanout_prefix': True,
        'fanout_patterns': True,
    }
    CELERY_RESULT_BACKEND = 'redis://localhost:6379'
    CELERY_ACCEPT_CONTENT = ['pickle', 'json']
    CELERYD_HIJACK_ROOT_LOGGER = False
    CELERYBEAT_SCHEDULER = 'udata.tasks.Scheduler'
    CELERY_MONGODB_SCHEDULER_COLLECTION = "schedules"
    # CELERY_TASK_SERIALIZER = 'pickle'
    # CELERYD_POOL = 'gevent'

    CACHE_KEY_PREFIX = 'udata-cache'
    CACHE_TYPE = 'redis'

    SECURITY_PASSWORD_HASH = b'bcrypt'
    SECURITY_PASSWORD_SALT = b'udata'

    OAUTH2_PROVIDER_TOKEN_EXPIRES_IN = 30 * 24 * 60 * 60  # 30 days

    MAIL_DEFAULT_SENDER = 'webmaster@udata'

    AUTO_INDEX = True

    SITE_ID = 'default'
    SITE_TITLE = 'uData'
    SITE_KEYWORDS = ['opendata', 'udata']
    SITE_AUTHOR_URL = None
    SITE_AUTHOR = None
    SITE_GITHUB_URL = 'https://github.com/etalab/udata'
    USE_SSL = False

    PLUGINS = []
    THEME = 'default'

    STATIC_DIRS = []

    OAUTH2_PROVIDER_ERROR_ENDPOINT = 'oauth-i18n.oauth_error'

    MD_ALLOWED_TAGS = [
        'a',
        'abbr',
        'acronym',
        'b',
        'br',
        'blockquote',
        'code',
        'dl',
        'dt',
        'em',
        'i',
        'li',
        'ol',
        'pre',
        'small',
        'strong',
        'ul',
    ]

    MD_ALLOWED_ATTRIBUTES = {
        'a': ['href', 'title'],
        'abbr': ['title'],
        'acronym': ['title'],
    }

    MD_ALLOWED_STYLES = []

    # Cache duration for templates.
    TEMPLATE_CACHE_DURATION = 5  # Minutes.

    DELAY_BEFORE_REMINDER_NOTIFICATION = 30  # Days

    HARVEST_PREVIEW_MAX_ITEMS = 20

    # CROQUEMORT = {
    #     'url': 'http://localhost:8000',
    #     'delay': 1,
    #     'retry': 10,
    # }

    # PIWIK_ID = # Demo = 11, Prod = 1
    # PIWIK_URL = 'stats.data.gouv.fr'
    # PIWIK_AUTH = '<32-chars-auth-token-from-piwik>'
    # # All keys are required.
    # PIWIK_GOALS = {
    #     'NEW_DATASET': # Demo = 1, Prod = 7
    #     'NEW_REUSE': # Demo = 2, Prod = 6
    #     'NEW_FOLLOW': # Demo = 3, Prod = 3
    #     'SHARE': , # Demo = 4, Prod = ?
    #     'RESOURCE_DOWNLOAD': , # Demo = 5, Prod = ?
    #     'RESOURCE_REDIRECT': , # Demo = 6, Prod = ?
    # }
    # TRACKING_BLACKLIST = ['api.notifications', 'api.checkurl']  # Default: []


class Testing(object):
    '''Sane values for testing. Should be applied as override'''
    TESTING = True
    SEND_MAIL = False
    WTF_CSRF_ENABLED = False
    AUTO_INDEX = False
    CELERY_ALWAYS_EAGER = True
    TEST_WITH_PLUGINS = False
    PLUGINS = []
    TEST_WITH_THEME = False
    THEME = 'default'
    CACHE_TYPE = 'null'
    CACHE_NO_NULL_WARNING = True
    DEBUG_TOOLBAR = False
    SERVER_NAME = 'localhost'
    DEFAULT_LANGUAGE = 'en'


class Debug(Defaults):
    DEBUG = True
    SEND_MAIL = False
    DEBUG_TB_INTERCEPT_REDIRECTS = False
    DEBUG_TB_PANELS = (
        'flask.ext.debugtoolbar.panels.versions.VersionDebugPanel',
        'flask.ext.debugtoolbar.panels.timer.TimerDebugPanel',
        'flask.ext.debugtoolbar.panels.headers.HeaderDebugPanel',
        'flask.ext.debugtoolbar.panels.request_vars.RequestVarsDebugPanel',
        'flask.ext.debugtoolbar.panels.config_vars.ConfigVarsDebugPanel',
        'flask.ext.debugtoolbar.panels.template.TemplateDebugPanel',
        'flask.ext.debugtoolbar.panels.logger.LoggingPanel',
        'flask.ext.debugtoolbar.panels.profiler.ProfilerDebugPanel',
        'flask.ext.mongoengine.panels.MongoDebugPanel',
    )
    CACHE_TYPE = 'null'
    CACHE_NO_NULL_WARNING = True
